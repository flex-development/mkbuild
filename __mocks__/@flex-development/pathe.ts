/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
