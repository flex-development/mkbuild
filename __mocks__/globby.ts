/**
 * @file Mocks - globby
 * @module mocks/globby
 * @see https://github.com/sindresorhus/globby
 */

/**
 * [`globby`][1] module type.
 *
 * [1]: https://github.com/sindresorhus/globby
 */
type Actual = typeof import('globby')

/**
 * `globby` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('globby')

export const globby = vi.fn(actual.globby)
