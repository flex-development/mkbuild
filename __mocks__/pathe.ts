/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
