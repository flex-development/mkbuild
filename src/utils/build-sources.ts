/**
 * @file Utilities - buildSources
 * @module mkbuild/utils/buildSources
 */

import {
  EXT_DTS_REGEX,
  EXT_TS_REGEX,
  IGNORE_PATTERNS
} from '#src/config/constants'
import type { Entry, SourceFile } from '#src/interfaces'
import fse from 'fs-extra'
import { globby, type Options as GlobbyOptions } from 'globby'
import * as pathe from 'pathe'

/**
 * Returns an array of build sources for `entry`.
 *
 * @async
 *
 * @param {Entry} entry - Build entry object
 * @param {string[] | string} [pattern='**'] - Globs matching source files
 * @param {string[]} [ignore=IGNORE_PATTERNS] - Globs excluding source files
 * @param {GlobbyOptions['fs']} [fs=fse] - Custom  `fs` methods
 * @return {Promise<SourceFile[]>} Source file objects
 */
const buildSources = async (
  entry: Entry,
  pattern: string[] | string = '**',
  ignore: string[] = IGNORE_PATTERNS,
  fs: GlobbyOptions['fs'] = fse
): Promise<SourceFile[]> => {
  /**
   * Relative paths to source files.
   *
   * **Note**: Files are relative to `entry.source` when bundling is disbaled.
   * When enabled, files are relative to `pathe.parse(entry.source).root`.
   *
   * @var {string[]} files
   */
  let files: string[] = []

  if (entry.bundle) {
    const { root } = pathe.parse(entry.source)
    files = [entry.source.replace(root + '/', '')]
  } else {
    files = await globby(pattern, {
      cwd: pathe.resolve(entry.absWorkingDir ?? process.cwd(), entry.source),
      dot: true,
      fs,
      ignore
    })
  }

  return files.map(file => {
    /**
     * {@link file} resolved.
     *
     * @const {string} path
     */
    const path: string = pathe.resolve(entry.source, entry.bundle ? '' : file)

    /**
     * {@link file} extension.
     *
     * @var {string} ext
     */
    let ext: string = pathe.extname(path)

    // ! fix extension if file is a typescript declaration file
    if (EXT_TS_REGEX.test(ext) && EXT_DTS_REGEX.test(file)) ext = '.d' + ext

    return { ext, file, path }
  })
}

export default buildSources
