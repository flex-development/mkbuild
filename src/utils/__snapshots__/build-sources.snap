// Vitest Snapshot v1

exports[`unit:utils/buildSources > should return build sources for bundle 1`] = `
[
  {
    "ext": ".ts",
    "file": "cli.ts",
  },
]
`;

exports[`unit:utils/buildSources > should return build sources for transpilation 1`] = `
[
  {
    "ext": ".ts",
    "file": "dbl-linear.ts",
  },
  {
    "ext": ".mts",
    "file": "reverse.mts",
  },
  {
    "ext": ".d.cts",
    "file": "save-mark.d.cts",
  },
]
`;
